README.md
# tts


## Overview
The `tts` ROS node enables a robot to speak with a human voice by providing a Text-To-Speech service.
Out of the box this package listens to a speech topic, submits text to the Amazon Polly cloud service to generate an audio stream file,
retrieves the audio stream from Amazon Polly, and plays the audio stream via the default output device.
The nodes can be configured to use different voices as well as custom lexicons and SSML tags which enable you to control aspects of speech,
such as pronunciation, volume, pitch, speed rate, etc. A [sample ROS application] with this node,
and more details on speech customization are available within the [Amazon Polly documentation].

**Amazon Polly Summary**: Amazon Polly is a service that turns text into lifelike speech, allowing you to create applications that talk,
and build entirely new categories of speech-enabled products. Amazon Polly is a Text-to-Speech service that uses advanced deep learning technologies to synthesize speech that sounds like a human voice.
With dozens of lifelike voices across a variety of languages, you can select the ideal voice and build speech-enabled applications that work in many different countries.

### License
The source code is released under an [Apache 2.0].

**Author**: AWS RoboMaker<br/>
**Affiliation**: [Amazon Web Services (AWS)]<br/>
**Maintainer**: AWS RoboMaker, ros-contributions@amazon.com

### Supported ROS Distributions
- Kinetic
- Melodic

### Build status
* Travis CI:
    * "master" branch [![Build Status](https://travis-ci.org/aws-robotics/tts-ros1.svg?branch=master)](https://travis-ci.org/aws-robotics/tts-ros1/branches)
    * "release-latest" branch [![Build Status](https://travis-ci.org/aws-robotics/tts-ros1.svg?branch=release-latest)](https://travis-ci.org/aws-robotics/tts-ros1/branches)
* ROS build farm:
    * ROS Kinetic @ u16.04 Xenial [![Build Status](http://build.ros.org/job/Kbin_uX64__tts__ubuntu_xenial_amd64__binary/badge/icon)](http://build.ros.org/job/Kbin_uX64__tts__ubuntu_xenial_amd64__binary)
    * ROS Melodic @ u18.04 Bionic [![Build Status](http://build.ros.org/job/Mbin_uB64__tts__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros.org/job/Mbin_uB64__tts__ubuntu_bionic_amd64__binary)


## Installation

### AWS Credentials
You will need to create an AWS Account and configure the credentials to be able to communicate with AWS services. You may find [AWS Configuration and Credential Files] helpful.

This node will require the following AWS account IAM role permissions:
- `polly:SynthesizeSpeech`

### Dependencies
In order to use the Text-To-Speech node with ROS kinetic you must update the version of boto3 that is installed on your system to at least version 1.9.0. You can do this by running the command:

        pip3 install -U boto3

This step is required before the node will work properly because the version of boto3 is not new enough for the features required by this node. 

### Binaries
On Ubuntu you can install the latest version of this package using the following command

        sudo apt-get update
        sudo apt-get install -y ros-$ROS_DISTRO-tts

### Building from Source

To build from source you'll need to create a new workspace, clone and checkout the latest release branch of this repository, install all the dependencies, and compile. If you need the latest development features you can clone from the `master` branch instead of the latest release branch. While we guarantee the release branches are stable, __the `master` should be considered to have an unstable build__ due to ongoing development. 

- Create a ROS workspace and a source directory

        mkdir -p ~/ros-workspace/src

- Clone the package into the source directory . 

        cd ~/ros-workspace/src
        git clone https://github.com/aws-robotics/tts-ros1.git -b release-latest

- Install dependencies

        cd ~/ros-workspace 
        sudo apt-get update && rosdep update
        rosdep install --from-paths src --ignore-src -r -y
        
_Note: If building the master branch instead of a release branch you may need to also checkout and build the master branches of the packages this package depends on._

- Build the packages

        cd ~/ros-workspace && colcon build

- Configure ROS library Path

        source ~/ros-workspace/install/setup.bash
        
- Build and run the unit tests

        colcon test --packages-select tts && colcon test-result --all

### Testing in Containers/Virtual Machines

Even if your container or virtual machine does not have audio device, you can still test TTS by leveraging an audio server.

The following is an example setup on a MacBook with PulseAudio as the audio server.
If you are new to PulseAudio, you may want to read the [PulseAudio Documentation].

**Step 1: Start PulseAudio on your laptop**

After installation, start the audio server with *module-native-protocol-tcp* loaded:

    pulseaudio --load=module-native-protocol-tcp --exit-idle-time=-1 --log-target=stderr -v

Note the extra arguments `-v` and `--log-target` are used for easier troubleshooting.

**Step 2: Run TTS nodes in container**

In your container, make sure you set the right environment variables.
For example, you can start the container using `docker run -it -e PULSE_SERVER=docker.for.mac.localhost ubuntu:16.04`.

Then you will be able to run ROS nodes in the container and hear the audio from your laptop speakers.

**Troubleshooting**

If your laptop has multiple audio output devices, make sure the right one has the right volume.
This command will give you a list of output devices and tell you which one has been selected:

    pacmd list-sinks | grep -E '(index:|name:|product.name)'

## Launch Files
An example launch file called `sample_application.launch` is provided.


## Usage

### Run the node
- **Plain text**
  - `roslaunch tts sample_application.launch`
  - `rosrun tts voicer.py 'Hello World'`

- **SSML**
  - `roslaunch tts sample_application.launch`
  - `rosrun tts voicer.py '<speak>Mary has a <amazon:effect name="whispered">little lamb.</amazon:effect></speak>' '{"text_type":"ssml"}'`


## Configuration File and Parameters
| Parameter Name | Type | Description |
| -------------- | ---- | ----------- |
| polly_action | *string* | Currently only one action named `SynthesizeSpeech` is supported. |
| text | *string* | The text to be synthesized. It can be plain text or SSML. See also `text_type`. |
| text_type | *string* | A user can choose from `text` and `ssml`. Default: `text`. |
| voice_id | *string* | The list of supported voices can be found on [official Amazon Polly document]. Default: Joanna |
| output_format | *string* | Valid formats are `ogg_vorbis`, `mp3` and `pcm`. Default: `ogg_vorbis` |
| output_path | *string* | The audio data will be saved as a local file for playback and reuse/inspection purposes. This parameter is to provide a preferred path to save the file. Default: `.` |
| sample_rate | *string* | Note `16000` is a valid sample rate for all supported formats. Default: `16000`. |


## Performance and Benchmark Results
We evaluated the performance of this node by runnning the followning scenario on a Raspberry Pi 3 Model B:
- Launch a baseline graph containing the talker and listener nodes from the [roscpp_tutorials package](https://wiki.ros.org/roscpp_tutorials), plus two additional nodes that collect CPU and memory usage statistics. Allow the nodes to run for 60 seconds. 
- Launch the nodes `polly_node`, `synthesizer_node` and `tts_node` by using the launch file `sample_application.launch` as described above. At the same time, perform several calls to the action `tts/action/Speech.action` using the `voicer.py` script descried above, by running the following script in the background:

```bash
rosrun tts voicer.py  '<speak>Amazon Polly is a <emphasis level="strong">Text-to-Speech</emphasis> (TTS) cloud service</speak>' '{"text_type":"ssml"}' ; sleep 1
rosrun tts voicer.py  '<speak>that converts text into lifelike speech</speak>' '{"text_type":"ssml"}' ; sleep 1
rosrun tts voicer.py  '<speak>You can use Amazon Polly to develop applications that increase <emphasis level="moderate">engagement and accessibility</emphasis></speak>' '{"text_type":"ssml"}' ; sleep 1
rosrun tts voicer.py  '<speak>Amazon Polly supports multiple languages and includes a variety of lifelike voices</speak>' '{"text_type":"ssml"}' ; sleep 1
rosrun tts voicer.py  '<speak>so you can build speech-enabled applications that work in multiple locations</speak>' '{"text_type":"ssml"}' ; sleep 1
rosrun tts voicer.py  '<speak>and use the ideal voice for your customers</speak>' '{"text_type":"ssml"}' ; sleep 1
```

- Allow the nodes to run for 180 seconds. 
- Terminate the  `polly_node`, `synthesizer_node` and `tts_node` nodes, and allow the reamaining nodes to run for 60 seconds.

The following graph shows the CPU usage during that scenario.  The 1 minute average CPU usage starts at 16.75% during the launch of the baseline graph, and stabilizes at 6%. When we launch the Polly nodes around second 85, the 1 minute average CPU increases up to a peak of 22.25% and stabilizes around 20%. After we stop making requests with the script `voicer.py` around second 206 the 1 minute average CPU usage moves to around 12%, and decreases gradually, and goes down again to 2.5 % after we stop the Polly nodes at the end of the scenario. 

![cpu](wiki/images/cpu.svg)

The following graph shows the memory usage during that scenario. We start with a memory usage of around 227 MB that increases to around 335 MB (+47.58%) when we lanch the Polly nodes around second 85, and gets to a peak of 361 MB (+59% wrt. initial value) while we are calling the script `voicer.py`. The memory usage goes back to the initial values after stopping the Polly nodes. 

![memory](wiki/images/memory.svg)


## Nodes

### polly
Polly node is the engine for the synthesizing job. It provides user-friendly yet powerful APIs so a user doesn't have to deal with technical details of AWS service calls.

#### Services
- **`polly (tts/Polly)`**

  Call the service to use Amazon Polly to synthesize the audio.

#### Reserved for future usage
- `language_code (string, default: None)`
  
  A user doesn't have to provide a language code and this is reserved for future usage.

- `lexicon_content (string, default: None)`

- `lexicon_name (string, default: None)`

- `lexicon_names (string[], default: empty)`

- `speech_mark_types (string[], default: empty)`

- `max_results (uint32, default: None)`

- `next_token (string, default: None)`

- `sns_topic_arn (string, default: None)`

- `task_id (string, default: None)`

- `task_status (string, default: iNone)`

- `output_s3_bucket_name (string, default: None)`

- `output_s3_key_prefix (string, default: None)`

- `include_additional_language_codes (bool, default: None)`

### synthesizer node

#### Services
- **`synthesizer (tts/Synthesizer)`**

  Call the service to synthesize.

#### Parameters

- **`text (string)`**

    The text to be synthesized.

- **`metadata (string, JSON format)`**

    Optional, for user to have control over how synthesis happens.

### tts node

#### Action

- **`speech`**

#### Parameters

- **`text (string)`**

  The text to be synthesized.

- **`metadata (string, JSON format)`**

  Optional, for user to have control over how synthesis happens.


## Bugs & Feature Requests
Please contact the team directly if you would like to request a feature.

Please report bugs in [Issue Tracker].


[AWS Configuration and Credential Files]: https://docs.aws.amazon.com/cli/latest/userguide/cli-config-files.html
[Amazon Polly documentation]: https://docs.aws.amazon.com/polly/latest/dg/what-is.html
[Amazon Web Services (AWS)]: https://aws.amazon.com/
[Apache 2.0]: https://aws.amazon.com/apache-2-0/
[Issue Tracker]: https://github.com/aws-robotics/tts-ros1/issues
[PulseAudio Documentation]: https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/
[official Amazon Polly document]: https://docs.aws.amazon.com/polly/latest/dg/voicelist.html
[sample ROS application]: https://github.com/aws-robotics/aws-robomaker-sample-application-voiceinteraction


tts/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package tts
^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2019-03-20)
------------------
* Merge pull request `#2 <https://github.com/aws-robotics/tts-ros1/issues/2>`_ from yyu/fix
  no assert_called() for older versions of mock
* no assert_called() for older versions of mock
* remove rostest from top level find_package (`#1 <https://github.com/aws-robotics/tts-ros1/issues/1>`_)
  It's conditionally found in the testing section only so it's only a test_depend
* Contributors: Tully Foote, Yuan "Forrest" Yu, y²

Forthcoming
-----------
* update changelog to be compatible with catkin_generate_changelog (`#28 <https://github.com/aws-robotics/tts-ros1/issues/28>`_)
  Signed-off-by: y² <forresty@amazon.com>
* increment patch version (`#27 <https://github.com/aws-robotics/tts-ros1/issues/27>`_)
  Signed-off-by: y² <forresty@amazon.com>
* Release 1.0.1 (`#9 <https://github.com/aws-robotics/tts-ros1/issues/9>`_)
  * Release 1.0.1
  * 1.0.1
* Merge pull request `#2 <https://github.com/aws-robotics/tts-ros1/issues/2>`_ from yyu/fix
  no assert_called() for older versions of mock
* no assert_called() for older versions of mock
* remove rostest from top level find_package (`#1 <https://github.com/aws-robotics/tts-ros1/issues/1>`_)
  It's conditionally found in the testing section only so it's only a test_depend
* Contributors: AAlon, M. M, Tully Foote, Yuan "Forrest" Yu, y²

1.0.0 (2019-03-20)
------------------


tts/test/test_unit_polly.py
#!/usr/bin/env python

# Copyright (c) 2018, Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#  http://aws.amazon.com/apache2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

from __future__ import print_function


from mock import patch, MagicMock # python2 uses backport of unittest.mock(docs.python.org/3/library/unittest.mock.html)
import unittest


class TestPolly(unittest.TestCase):

    def setUp(self):
        """important: import tts which is a relay package::

            devel/lib/python2.7/dist-packages/
            +-- tts
            |   +-- __init__.py
            +-- ...

        per http://docs.ros.org/api/catkin/html/user_guide/setup_dot_py.html:

        A relay package is a folder with an __init__.py folder and nothing else.
        Importing this folder in python will execute the contents of __init__.py,
        which will in turn import the original python modules in the folder in
        the sourcespace using the python exec() function.
        """
        import tts
        self.assertIsNotNone(tts)

    @patch('tts.amazonpolly.Session')
    def test_init(self, boto3_session_class_mock):
        from tts.amazonpolly import AmazonPolly
        AmazonPolly()

        boto3_session_class_mock.assert_called()
        boto3_session_class_mock.return_value.client.assert_called_with('polly')

    @patch('tts.amazonpolly.Session')
    def test_defaults(self, boto3_session_class_mock):
        from tts.amazonpolly import AmazonPolly
        polly = AmazonPolly()

        boto3_session_class_mock.assert_called()
        boto3_session_class_mock.return_value.client.assert_called_with('polly')

        self.assertEqual('text', polly.default_text_type)
        self.assertEqual('ogg_vorbis', polly.default_output_format)
        self.assertEqual('Joanna', polly.default_voice_id)
        self.assertEqual('.', polly.default_output_folder)
        self.assertEqual('output', polly.default_output_file_basename)

    @patch('tts.amazonpolly.Session')
    def test_good_synthesis_with_default_args(self, boto3_session_class_mock):
        boto3_session_obj_mock = MagicMock()
        boto3_polly_obj_mock = MagicMock()
        boto3_polly_response_mock = MagicMock()
        audio_stream_mock = MagicMock()
        fake_audio_stream_data = 'I am audio.'
        fake_audio_content_type = 'super tts'
        fake_boto3_polly_response_metadata = {'foo': 'bar'}

        boto3_session_class_mock.return_value = boto3_session_obj_mock
        boto3_session_obj_mock.client.return_value = boto3_polly_obj_mock
        boto3_polly_obj_mock.synthesize_speech.return_value = boto3_polly_response_mock
        audio_stream_mock.read.return_value = fake_audio_stream_data
        d = {
            'AudioStream': audio_stream_mock,
            'ContentType': fake_audio_content_type,
            'ResponseMetadata': fake_boto3_polly_response_metadata
        }
        boto3_polly_response_mock.__contains__.side_effect = d.__contains__
        boto3_polly_response_mock.__getitem__.side_effect = d.__getitem__

        from tts.amazonpolly import AmazonPolly
        polly_under_test = AmazonPolly()

        boto3_session_class_mock.assert_called()
        boto3_session_obj_mock.client.assert_called_with('polly')

        res = polly_under_test.synthesize(text='hello')

        expected_synthesize_speech_kwargs = {
            'LexiconNames': [],
            'OutputFormat': 'ogg_vorbis',
            'SampleRate': '22050',
            'SpeechMarkTypes': [],
            'Text': 'hello',
            'TextType': 'text',
            'VoiceId': 'Joanna',
        }
        boto3_polly_obj_mock.synthesize_speech.assert_called_with(**expected_synthesize_speech_kwargs)

        from tts.srv import PollyResponse
        self.assertTrue(isinstance(res, PollyResponse))

        import json
        j = json.loads(res.result)
        observed_audio_file_content = open(j['Audio File']).read()
        self.assertEqual(fake_audio_stream_data, observed_audio_file_content)

        self.assertEqual(fake_audio_content_type, j['Audio Type'])
        self.assertEqual(str(fake_boto3_polly_response_metadata), j['Amazon Polly Response Metadata'])

    @patch('tts.amazonpolly.Session')
    def test_polly_raises(self, boto3_session_class_mock):
        boto3_session_obj_mock = MagicMock()
        boto3_polly_obj_mock = MagicMock()
        boto3_polly_response_mock = MagicMock()
        audio_stream_mock = MagicMock()
        fake_audio_stream_data = 'I am audio.'
        fake_audio_content_type = 'super voice'
        fake_boto3_polly_response_metadata = {'foo': 'bar'}

        boto3_session_class_mock.return_value = boto3_session_obj_mock
        boto3_session_obj_mock.client.return_value = boto3_polly_obj_mock
        boto3_polly_obj_mock.synthesize_speech.side_effect = RuntimeError('Amazon Polly Exception')
        audio_stream_mock.read.return_value = fake_audio_stream_data
        d = {
            'AudioStream': audio_stream_mock,
            'ContentType': fake_audio_content_type,
            'ResponseMetadata': fake_boto3_polly_response_metadata
        }
        boto3_polly_response_mock.__contains__.side_effect = d.__contains__
        boto3_polly_response_mock.__getitem__.side_effect = d.__getitem__

        from tts.amazonpolly import AmazonPolly
        polly_under_test = AmazonPolly()

        boto3_session_class_mock.assert_called()
        boto3_session_obj_mock.client.assert_called_with('polly')

        res = polly_under_test.synthesize(text='hello')

        expected_synthesize_speech_kwargs = {
            'LexiconNames': [],
            'OutputFormat': 'ogg_vorbis',
            'SampleRate': '22050',
            'SpeechMarkTypes': [],
            'Text': 'hello',
            'TextType': 'text',
            'VoiceId': 'Joanna',
        }
        boto3_polly_obj_mock.synthesize_speech.assert_called_with(**expected_synthesize_speech_kwargs)

        from tts.srv import PollyResponse
        self.assertTrue(isinstance(res, PollyResponse))

        import json
        j = json.loads(res.result)
        self.assertTrue('Exception' in j)
        self.assertTrue('Traceback' in j)

    @patch('tts.amazonpolly.AmazonPolly')
    def test_cli(self, amazon_polly_class_mock):
        import sys
        with patch.object(sys, 'argv', ['polly_node.py', '-n', 'polly-node']):
            from tts import amazonpolly
            amazonpolly.main()
            amazon_polly_class_mock.assert_called()
            amazon_polly_class_mock.return_value.start.assert_called_with(node_name='polly-node', service_name='polly')


if __name__ == '__main__':
    import rosunit
    rosunit.unitrun('tts', 'unittest-polly', TestPolly)


tts/test/test_unit_synthesizer.py
#!/usr/bin/env python

# Copyright (c) 2018, Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#  http://aws.amazon.com/apache2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

from __future__ import print_function

from mock import patch, MagicMock # python2 uses backport of unittest.mock(docs.python.org/3/library/unittest.mock.html)
import unittest


class TestSynthesizer(unittest.TestCase):

    def setUp(self):
        """important: import tts which is a relay package::

            devel/lib/python2.7/dist-packages/
            +-- tts
            |   +-- __init__.py
            +-- ...

        per http://docs.ros.org/api/catkin/html/user_guide/setup_dot_py.html:

        A relay package is a folder with an __init__.py folder and nothing else.
        Importing this folder in python will execute the contents of __init__.py,
        which will in turn import the original python modules in the folder in
        the sourcespace using the python exec() function.
        """
        import tts
        self.assertIsNotNone(tts)

    def test_init(self):
        from tts.synthesizer import SpeechSynthesizer
        speech_synthesizer = SpeechSynthesizer()
        self.assertEqual('text', speech_synthesizer.default_text_type)

    @patch('tts.amazonpolly.AmazonPolly')
    def test_good_synthesis_with_mostly_default_args_using_polly_lib(self, polly_class_mock):
        polly_obj_mock = MagicMock()
        polly_class_mock.return_value = polly_obj_mock

        test_text = 'hello'
        test_metadata = '''
            {
                "output_path": "/tmp/test"
            }
        '''
        expected_polly_synthesize_args = {
            'output_format': 'ogg_vorbis',
            'voice_id': 'Joanna',
            'sample_rate': '22050',
            'text_type': 'text',
            'text': test_text,
            'output_path': "/tmp/test"
        }

        from tts.synthesizer import SpeechSynthesizer
        from tts.srv import SynthesizerRequest
        speech_synthesizer = SpeechSynthesizer(engine='POLLY_LIBRARY')
        request = SynthesizerRequest(text=test_text, metadata=test_metadata)
        response = speech_synthesizer._node_request_handler(request)

        polly_class_mock.assert_called()
        polly_obj_mock.synthesize.assert_called_with(**expected_polly_synthesize_args)

        self.assertEqual(response.result, polly_obj_mock.synthesize.return_value.result)

    @patch('tts.amazonpolly.AmazonPolly')
    def test_synthesis_with_bad_metadata_using_polly_lib(self, polly_class_mock):
        polly_obj_mock = MagicMock()
        polly_class_mock.return_value = polly_obj_mock

        test_text = 'hello'
        test_metadata = '''I am no JSON'''

        from tts.synthesizer import SpeechSynthesizer
        from tts.srv import SynthesizerRequest
        speech_synthesizer = SpeechSynthesizer(engine='POLLY_LIBRARY')
        request = SynthesizerRequest(text=test_text, metadata=test_metadata)
        response = speech_synthesizer._node_request_handler(request)

        self.assertTrue(response.result.startswith('Exception: '))

    @patch('tts.amazonpolly.AmazonPolly')
    def test_bad_engine(self, polly_class_mock):
        polly_obj_mock = MagicMock()
        polly_class_mock.return_value = polly_obj_mock

        ex = None

        from tts.synthesizer import SpeechSynthesizer
        try:
            SpeechSynthesizer(engine='NON-EXIST ENGINE')
        except Exception as e:
            ex = e

        self.assertTrue(isinstance(ex, SpeechSynthesizer.BadEngineError))

    def test_cli_help_message(self):
        import os
        source_file_dir = os.path.dirname(os.path.abspath(__file__))
        synthersizer_path = os.path.join(source_file_dir, '..', 'scripts', 'synthesizer_node.py')
        import subprocess
        o = subprocess.check_output(['python', synthersizer_path, '-h'])
        self.assertTrue(str(o).startswith('Usage: '))

    @patch('tts.synthesizer.SpeechSynthesizer')
    def test_cli_engine_dispatching_1(self, speech_synthesizer_class_mock):
        import sys
        with patch.object(sys, 'argv', ['synthesizer_node.py']):
            import tts.synthesizer
            tts.synthesizer.main()
            speech_synthesizer_class_mock.assert_called_with(engine='POLLY_SERVICE', polly_service_name='polly')
            speech_synthesizer_class_mock.return_value.start.assert_called_with(node_name='synthesizer_node',
                                                                               service_name='synthesizer')

    @patch('tts.synthesizer.SpeechSynthesizer')
    def test_cli_engine_dispatching_2(self, speech_synthesizer_class_mock):
        import sys
        with patch.object(sys, 'argv', ['synthesizer_node.py', '-e', 'POLLY_LIBRARY']):
            from tts import synthesizer
            synthesizer.main()
            speech_synthesizer_class_mock.assert_called_with(engine='POLLY_LIBRARY')
            speech_synthesizer_class_mock.return_value.start.assert_called()

    @patch('tts.synthesizer.SpeechSynthesizer')
    def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
        import sys
        with patch.object(sys, 'argv', ['synthesizer_node.py', '-p', 'apolly']):
            from tts import synthesizer
            synthesizer.main()
            speech_synthesizer_class_mock.assert_called_with(engine='POLLY_SERVICE', polly_service_name='apolly')
            speech_synthesizer_class_mock.return_value.start.assert_called()


if __name__ == '__main__':
    import rosunit
    rosunit.unitrun('tts', 'unittest-synthesizer', TestSynthesizer)


tts/src/tts/synthesizer.py
#!/usr/bin/env python

# Copyright (c) 2018, Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#  http://aws.amazon.com/apache2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

import os
import time
import json
import rospy
import hashlib
from optparse import OptionParser
from tts.srv import Synthesizer, SynthesizerResponse


class SpeechSynthesizer:
    """This class serves as a ROS service node that should be an entry point of a TTS task.

    Although the current implementation uses Amazon Polly as the synthesis engine, it is not hard to let it support
    more heterogeneous engines while keeping the API the same.

    In order to support a variety of engines, the SynthesizerRequest was designed with flexibility in mind. It
    has two fields: text and metadata. Both are strings. In most cases, a user can ignore the metadata and call
    the service with some plain text. If the use case needs any control or engine-specific feature, the extra
    information can be put into the JSON-form metadata. This class will use the information when calling the engine.

    The decoupling of the synthesizer and the actual synthesis engine will benefit the users in many ways.

    First, a user will be able to use a unified interface to do the TTS job and have the freedom to use different
    engines available with no or very little change from the client side.

    Second, by applying some design patterns, the synthesizer can choose an engine dynamically. For example, a user
    may prefer to use Amazon Polly but is also OK with an offline solution when network is not reliable.

    Third, engines can be complicated, thus difficult to use. As an example, Amazon Polly supports dozens of parameters
    and is able to accomplish nontrivial synthesis jobs, but majority of the users never need those features. This
    class provides a clean interface with two parameters only, so that it is much easier and pleasant to use. If by
    any chance the advanced features are required, the user can always leverage the metadata field or even go to the
    backend engine directly.

    Also, from an engineering perspective, simple and decoupled modules are easier to maintain.

    This class supports two modes of using polly. It can either call a service node or use AmazonPolly as a library.

    Start the service node::

        $ rosrun tts synthesizer_node.py  # use default configuration
        $ rosrun tts synthesizer_node.py -e POLLY_LIBRARY  # will not call polly service node

    Call the service::

        $ rosservice call /synthesizer 'hello' ''
        $ rosservice call /synthesizer '<speak>hello</speak>' '"{\"text_type\":\"ssml\"}"'
    """

    class PollyViaNode:
        def __init__(self, polly_service_name='polly'):
            self.service_name = polly_service_name

        def __call__(self, **kwargs):
            rospy.loginfo('will call service {}'.format(self.service_name))
            from tts.srv import Polly
            rospy.wait_for_service(self.service_name)
            polly = rospy.ServiceProxy(self.service_name, Polly)
            return polly(polly_action='SynthesizeSpeech', **kwargs)

    class PollyDirect:
        def __init__(self):
            pass

        def __call__(self, **kwargs):
            rospy.loginfo('will import amazonpolly.AmazonPolly')
            from tts.amazonpolly import AmazonPolly
            node = AmazonPolly()
            return node.synthesize(**kwargs)

    ENGINES = {
        'POLLY_SERVICE': PollyViaNode,
        'POLLY_LIBRARY': PollyDirect,
    }

    class BadEngineError(NameError):
        pass

    def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
        if engine not in self.ENGINES:
            msg = 'bad engine {} which is not one of {}'.format(engine, ', '.join(SpeechSynthesizer.ENGINES.keys()))
            raise SpeechSynthesizer.BadEngineError(msg)

        engine_kwargs = {'polly_service_name': polly_service_name} if engine == 'POLLY_SERVICE' else {}
        self.engine = self.ENGINES[engine](**engine_kwargs)

        self.default_text_type = 'text'
        self.default_voice_id = 'Joanna'
        self.default_output_format = 'ogg_vorbis'

    def _call_engine(self, **kw):
        """Call engine to do the job.

        If no output path is found from input, the audio file will be put into /tmp and the file name will have
        a prefix of the md5 hash of the text.

        :param kw: what AmazonPolly needs to synthesize
        :return: response from AmazonPolly
        """
        if 'output_path' not in kw:
            tmp_filename = hashlib.md5(kw['text']).hexdigest()
            tmp_filepath = os.path.join(os.sep, 'tmp', 'voice_{}_{}'.format(tmp_filename, str(time.time())))
            kw['output_path'] = os.path.abspath(tmp_filepath)
        rospy.loginfo('audio will be saved as {}'.format(kw['output_path']))

        return self.engine(**kw)

    def _parse_request_or_raise(self, request):
        """It will raise if request is malformed.

        :param request: an instance of SynthesizerRequest
        :return: a dict
        """
        md = json.loads(request.metadata) if request.metadata else {}

        md['output_format'] = md.get('output_format', self.default_output_format)
        md['voice_id'] = md.get('voice_id', self.default_voice_id)
        md['sample_rate'] = md.get('sample_rate', '16000' if md['output_format'].lower() == 'pcm' else '22050')
        md['text_type'] = md.get('text_type', self.default_text_type)
        md['text'] = request.text

        return md

    def _node_request_handler(self, request):
        """The callback function for processing service request.

        It never raises. If anything unexpected happens, it will return a SynthesizerResponse with the exception.

        :param request: an instance of SynthesizerRequest
        :return: a SynthesizerResponse
        """
        rospy.loginfo(request)
        try:
            kws = self._parse_request_or_raise(request)
            res = self._call_engine(**kws).result

            return SynthesizerResponse(res)
        except Exception as e:
            return SynthesizerResponse('Exception: {}'.format(e))

    def start(self, node_name='synthesizer_node', service_name='synthesizer'):
        """The entry point of a ROS service node.

        :param node_name: name of ROS node
        :param service_name:  name of ROS service
        :return: it doesn't return
        """
        rospy.init_node(node_name)

        service = rospy.Service(service_name, Synthesizer, self._node_request_handler)

        rospy.loginfo('{} running: {}'.format(node_name, service.uri))

        rospy.spin()


def main():
    usage = '''usage: %prog [options]
    '''

    parser = OptionParser(usage)

    parser.add_option("-n", "--node-name", dest="node_name", default='synthesizer_node',
                      help="name of the ROS node",
                      metavar="NODE_NAME")
    parser.add_option("-s", "--service-name", dest="service_name", default='synthesizer',
                      help="name of the ROS service",
                      metavar="SERVICE_NAME")
    parser.add_option("-e", "--engine", dest="engine", default='POLLY_SERVICE',
                      help="name of the synthesis engine",
                      metavar="ENGINE")
    parser.add_option("-p", "--polly-service-name", dest="polly_service_name", default='polly',
                      help="name of the polly service",
                      metavar="POLLY_SERVICE_NAME")

    (options, args) = parser.parse_args()

    node_name = options.node_name
    service_name = options.service_name
    engine = options.engine
    polly_service_name = options.polly_service_name

    if engine == 'POLLY_SERVICE':
        synthesizer = SpeechSynthesizer(engine=engine, polly_service_name=polly_service_name)
    else:
        synthesizer = SpeechSynthesizer(engine=engine)
    synthesizer.start(node_name=node_name, service_name=service_name)


if __name__ == "__main__":
    main()
